src/tsolvers/egraph/EgraphSolver.cc

```diff
@@ -213,23 +215,131 @@ void Egraph::getConflict( bool deduction, vec<PtAsgn>& cnfl )
 
 void Egraph::clearModel()
 {
-    values.clear();
-    values_ok = false;
+    values.reset(nullptr);
 }
 
 void Egraph::computeModel( )
 {
-    if (values_ok == true)
+    if (values != nullptr)
         return;
 
-    const vec<ERef>& enodes = enode_store.getTermEnodes();
-    for (int i = 0; i < enodes.size(); i++) {
-        if (values.has(enodes[i]))
-            continue;
-        ERef root_r = enode_store[enodes[i]].getRoot();
-        values.insert(enodes[i], root_r);
+    values = std::unique_ptr<Values>(new Values());
+    for (ERef er : enode_store.getTermEnodes()) {
+        ERef root_r = enode_store[er].getRoot();
+        values->addValue(er, root_r);
+    }
+}
+
+/**
+ * Get the abstract value (name of an element of the universe) corresponding to er.
+ * If er is ERef_Undef, use the default element for the sort sr
+ * @param er
+ * @param sr the sort of er (used if er is ERef_Undef)
+ * @return
+ */
+PTRef Egraph::getAbstractValueForERef(ERef er, SRef sr) const {
+
+    if (er == ERef_Undef) {
+        return logic.getDefaultValuePTRef(sr);
+    }
+
+    ERef val_er = values->getValue(er);
+    PTRef val_tr = enode_store.getPTRef(val_er);
+
+    if (isConstant(val_er)) {
+        return val_tr;
     }
-    values_ok = true;
+    std::stringstream ss;
+    ss << Logic::s_abstract_value_prefix << values->getValueIndex(val_er);
+    return logic.mkConst(logic.getSortRef(val_tr), ss.str().c_str());
+}
+
+void Egraph::fillTheoryFunctions(ModelBuilder & modelBuilder, const MapWithKeys<PTRef,PTRef,PTRefHash>& substs) const
+{
+    Map<PTRef,bool,PTRefHash> seen_substs;
+    for (ERef er : enode_store.getTermEnodes()) {
+        PTRef tr = enode_store.getPTRef(er);
+
+
+        if (logic.hasSortBool(tr) and (not logic.isUP(tr))) {
+            continue; // The Boolean return sorted terms that are not uninterpreted predicates come from the SAT solver
+        }
+
+        // If tr was substituted, obtain the real value from the substitution target
+        PTRef target_tr{PTRef_Undef};
+        ERef target_er{ERef_Undef};
+        if (substs.peek(tr, target_tr)) {
+            seen_substs.insert(tr, true);
+            enode_store.peekERef(target_tr, target_er);
+            if (target_er == ERef_Undef) {
+                // Substitution removed target_tr from input instance, it was never introduced to the SAT solver, and
+                // therefore also theory solver is not aware of it.
+                // The original term is guaranteed to be in the solver and is authoritative in this case.
+                target_er = er;
+            }
+        } else {
+            assert(([&](PTRef tr) { ERef tmp; enode_store.peekERef(tr, tmp); return tmp == er; })(tr));
+            assert(enode_store.peekERef(tr, target_er));
+            assert(target_er == er);
```

Comment:
Actually, don't these two asserts say the same thing as the one with lambda?

Reply:
Yes, I agree.  The lambda looks a little nicer, at least it's everything in one place...